osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs

```diff
@@ -283,6 +291,23 @@ protected virtual void CheckForResult(bool userTriggered, double timeOffset)
         /// </summary>
         /// <param name="judgement">The <see cref="Judgement"/> that provides the scoring information.</param>
         protected virtual JudgementResult CreateResult(Judgement judgement) => new JudgementResult(judgement);
+
+        /// <summary>
+        /// Applies the failing animation on this <see cref="DrawableHitObject"/>.
+        /// </summary>
+        public virtual void Fail()
+        {
+            if (HasFailed)
```

Comment:
One thing we try to avoid is including constant conditions in the virtual portion of the method as this generally leads to failures (especially true because this is `Ruleset` code, which may be implemented by users that make their own rulesets)

If the use overrides Fail but doesn't block on `if (HasFailed)`, their code will be called every frame. The same applies if they forget to call `base.Fail()` in their override (which I'm guessing objects *will* do if they don't want to apply the default animation) single `HasFailed` won't be set to true.

Best to move the virtual/transform part to a protected virtual method called by `Fail()`

```csharp
		/// <summary>
        /// Applies the failing animation on this <see cref="DrawableHitObject"/>.
        /// </summary>
        public virtual void Fail()
        {
			if (HasFailed) return;
			OnFail();
			HasFailed = true;
		}

		protected virtual void OnFail()
		{
			// transforms
		}
```

Reply:
Nice one, never thought of that..